pkg/syncer/fullsync.go

```diff
@@ -234,14 +235,26 @@ func fullSyncDeleteVolumes(ctx context.Context, volumeIDDeleteArray []cnstypes.C
 		return
 	}
 	// Verify if Volume is not in use by any other Cluster before removing CNS tag
+	vcVersion, err := metadataSyncer.volumeManager.GetVCVersion(ctx)
+	if err != nil {
+		log.Errorf("FullSync: failed to get vc version. Err: %v", err)
+		return
+	}
 	for _, queryResult := range allQueryResults {
 		for _, volume := range queryResult.Volumes {
 			inUsebyOtherK8SCluster := false
-			for _, metadata := range volume.Metadata.EntityMetadata {
-				if metadata.(*cnstypes.CnsKubernetesEntityMetadata).ClusterID != metadataSyncer.configInfo.Cfg.Global.ClusterID {
+			if vcVersion == cns.ReleaseVSAN67u3 {
+				if volume.Metadata.ContainerCluster.ClusterId != metadataSyncer.configInfo.Cfg.Global.ClusterID {
```

Comment:
Why not set the EntityMetadata.ClusterId to volume.Metadata.ContainerCluster.ClusterId in govmomi if VC version is 6.7U3? This way we can only check EntityMetadata.ClusterId and avoid VC version specific logic in CSI.

Reply:
ClusterID is introduced in vSphere 7.0 and was not available in the vSphere 67 release.

Refer - https://vdc-download.vmware.com/vmwb-repository/dcr-public/8ed923df-bad4-49b3-b677-45bca5326e85/d2d90bb6-d1b3-4266-8ce5-443680187a9a/vim.cns.EntityMetadata.html

We should drop fields server is not understanding while marking a request, but we should not populate fields server has not returned. We are dropping unknown fields introduced in the newer releases to correct SOAP requests on older vSphere release, but on the response, we have received from the server we should not be altering anything.

Reply:
>we have received from the server we should not be altering anything
This is fine in govmomi. But we can modify volume manager to set EntityMetadata.ClusterId to volume.Metadata.ContainerCluster.ClusterId depending on the VC version. Basically we are leaking VC version specific logic in many places and it will be nice to consolidate VC version check to a single place like volume manager.

Reply:
> This is fine in govmomi.

I think we should treat govmomi as a generic SDK like pyvmomi. There are other clients, other than CSI driver for example govc command line support. dev may have used govmomi in their automation, so altering response from the server is not good.

> Basically we are leaking VC version specific logic in many places 

We are just using the VC version in the full sync only for query volume response. 
There was already one check in the controller to publish capability depending on VC version, we can not avoid that check.

![Screen Shot 2020-10-06 at 12 11 27 PM](https://user-images.githubusercontent.com/22985595/95249117-16f86600-07cd-11eb-870f-165acdf98d8c.png)



> But we can modify volume manager to set EntityMetadata.ClusterId to volume.Metadata.ContainerCluster.ClusterId depending on the VC version

I think this can slow down performance. We will need to traverse through elements in the array of CnsEntityMetadata[] and alter the object and set cluster-id before returning responce.